### PR TITLE
Add Superdense Coding to examples.

### PR DESCRIPTION
**Pull request by davemc84**
Very simplistic Superdense Coding example that might serve as an instructive file.

### CODE REVIEW

**Comment by davemc84** [comment]
I just signed the Contributor License Agreement!

File: examples/superdense_coding.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+"""Superdense Coding.
+Superdense Coding is a method to transmit two classical bits of information
+from a sender to a receiver by sending only one qubit. This is accomplished
+by pre-sharing an entangled qubit.
+
+The follwoing example sets qubit 0 to 0 and qubit 1 to 1 and are output in
```
Comment (Ashalynd) [inline examples/superdense_coding.py]: Typo: c/follwoing/following

File: examples/superdense_coding.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+1=1
+3=0
+4=1
+
+"""
+
+import cirq
+
+
+def make_superdense_circuit():
```
Comment (Ashalynd) [inline examples/superdense_coding.py]: Could you add comments to explain how the steps in your implementation relate to the stages in superdense coding process?

File: examples/superdense_coding.py (hunk 1 above)
Comment (davemc84) [reply examples/superdense_coding.py]: Thanks so much for your thoughts here - I really appreciate it!

I just updated the code for this typo as well as the commentary point.

File: examples/superdense_coding.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+"""Superdense Coding.
+Superdense Coding is a method to transmit two classical bits of information
+from a sender to a receiver by sending only one qubit. This is accomplished
+by pre-sharing an entangled qubit.
+
+The following example sets qubit 0 to 0 and qubit 1 to 1 and are output in
+qubits 3 and 4.
```
Comment (Ashalynd) [inline examples/superdense_coding.py]: This whole phrase is difficult to parse :)

**Comment by Ashalynd** [comment]
Thanks, this looks about right to me. To make this example a bit more interesting, could you add a possibility to set random combination of input bits? See e.g how it was done in Cirq/examples/grover.py.

File: examples/superdense_coding.py (hunk 3 above)
Comment (davemc84) [reply examples/superdense_coding.py]: Thanks very much for the feedback!  I just revised the language to hopefully make it easier to understand. I also agree with additional comment on making the input bits random and will get started on that right away.

Thanks again!

**Comment by davemc84** [comment]
Thanks again for the comments! I just updated the code to allow for the input bits to be randomly set to either 0 or 1.

Please let me know if there are any further edits that would be helpful to make here.

**Comment by Ashalynd** [comment]
Thanks for the quick updates! Please make sure that all the automatic checks pass, then you should be good to go.

**Comment by davemc84** [comment]
Thanks so much @Ashalynd for all of your help and your prompt feedback! Apologies for bugging you with this, but I am having some issues with clearing the ["pytest-and-incremental-coverage"](https://github.com/quantumlib/Cirq/pull/1418/checks?sha=1f9b0f976b77e655f72b26c05b87d77c6f713608) test. I have tried to troubleshoot what is causing this, but have had no luck, unfortunately. Is there a resource that you might be able to point me to that could help me to rectify this issue?

Thanks again for your help and apologies for bugging you with this.

**Comment by vtomole** [comment]
@davemc84 Please call this program in `examples/examples_test.py`.

**Comment by Ashalynd** [comment]
Travis shows the following message:

```
************* examples/superdense_coding.py (1 uncovered)
Line   50 is new or changed but not covered:          circuit.append([cirq.X(q0)])
```

TBH I am not sure why it complains about this line only, it doesn't make sense to me.
I am OK with skipping coverage for that line.

**Comment by davemc84** [comment]
Thanks so much @Ashalynd and @vtomole for all of the help and feedback!  Is there anything else that I can do on this?

Thanks again!

**Comment by Ashalynd** [comment]
> Thanks so much @Ashalynd and @vtomole for all of the help and feedback! Is there anything else that I can do on this?

Try disabling test coverage for this line and see if it makes Travis happy? Otherwise the issue has been somewhere else.

**Comment by vtomole** [comment]
`if random.randint(0, 1):` only runs depending on the output of `randint()`. That line should skipped to get passed the coverage.

**Comment by davemc84** [comment]
Thanks again for the help here! Apologies if this is a rudimentary question, but how would I go about skipping coverage for that line of code?

`circuit.append([cirq.X(q0)])`

After a bit of research, I was unable to find an answer to this, unfortunately.

**Comment by vtomole** [comment]
@davemc84  Add `# coverage: ignore` next to the line you want ignored.

**Comment by vtomole** [comment]
@davemc84 Add for the other `if` statement too.

**Comment by davemc84** [comment]
Thanks again, @Ashalynd and @vtomole!  Really do appreciate all of the help and guidance.  Please do let me know if there are any next steps or if there is anything else that I can do to be helpful here.

File: examples/superdense_coding.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+"""
+
+import random
+import cirq
+
+
+def make_superdense_circuit():
+    circuit = cirq.Circuit()
+    (q0, q1, q2, q3, q4) = cirq.LineQubit.range(5)
```
Comment (Strilanc) [inline examples/superdense_coding.py]: Nit: unnecessary parens

File: examples/superdense_coding.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    # Randomly sets q0 and q1 to either 0 or 1
+    if random.randint(0, 1):
+        circuit.append([cirq.X(q0)]) # coverage: ignore
```
Comment (Strilanc) [inline examples/superdense_coding.py]: You don't need the square brackets when appending single items.

File: examples/superdense_coding.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    if random.randint(0, 1):
```
Comment (Strilanc) [inline examples/superdense_coding.py]: How about embedding the randomness into the circuit itself? The gate sequence Hadamard Measure will produce a random bit. We could also use a depolarizing channel to randomize the qubit.

Comment (davemc84) [reply examples/superdense_coding.py]: Hi @Strilanc, thanks so much for these thoughts.  I really do appreciate you passing them along! I should have some time to incorporate a bit later today. By the way, I am a huge fan of [your website](http://algassert.com). Thanks so much for your work on that as well!

Comment (Strilanc) [reply examples/superdense_coding.py]: You could use one measurement for the inputs `measure(in1, in2, key="input")` and one for the outputs, and then print out the results in a way that made it obvious they agreed.

File: examples/superdense_coding.py (hunk 4 above)
Comment (davemc84) [reply examples/superdense_coding.py]: Got it - just made that tweak.

File: examples/superdense_coding.py (hunk 5 above)
Comment (davemc84) [reply examples/superdense_coding.py]: Thanks very much for the heads-up here. That line has now been removed given the change to how these are randomized, though very good to know for the future!

File: examples/superdense_coding.py (hunk 6 above)
Comment (davemc84) [reply examples/superdense_coding.py]: Great suggestion! I just reflected this change in the latest commit, though am still trying to better understand how to parse the TrialResult, so there is probably some further clean-up that can be made down the road.  That said, please do let me know if there are any other suggestions on how to improve this.

Thanks again for all of your help!!!

File: examples/superdense_coding.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    return circuit
+
+
+def main():
+    circuit = make_superdense_circuit()
+    print("Circuit:")
+    print(circuit)
+
+    sim = cirq.Simulator()
+    results = sim.run(circuit)
```
Comment (Strilanc) [inline examples/superdense_coding.py]: Use `repetitions=20` or similar, so you get many samples

Comment (davemc84) [reply examples/superdense_coding.py]: Thanks so much for this thought! I just incorporated into the latest commit.

Please feel free to pass along any additional changes that you might have and I would be more than happy to incorporate!

Comment (davemc84) [reply examples/superdense_coding.py]: Hi @Strilanc - my apologies for bugging you here, but please let me know if there are any further edits that you would like for me to make here or do to be helpful.  Thanks again for your thoughtful suggestions!!!

**Comment by Strilanc** [comment]
Looks good!

**Pull request merged**